### PR TITLE
Tor hidden service builder + more logs + tor example

### LINKED ISSUE #1100: Integration with the Tor control port
Create an (incomplete) rust client for the tor control port. This is needed to enable seamless tor integration on mobile and desktop.

This should enable the application to (TODO: create issues):
- Create a new ephemeral/permanent onion address on-the-fly
- Delete an onion address 
- Fetch the socks bind address for a given onion address (if the user pre-configures an onion address)

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

- Added HiddenServiceBuilder to setup a new or existing Tor Hidden Service
- Allow comms to configure itself (specifically, configure itself to use
      the SOCKS5 proxy) to use a given Hidden Service.
- The private key for the hidden service could be persisted and re-used
      later. The implementation for that is left to client code.
- Added `GETINFO` support to tor control port client
- Added a tor ping pong example.
- Had to change `Transport` trait and `TcpTransport` to get around Rust ICE 

```rust
    let tor_hidden_service = tor::HiddenServiceBuilder::new()
        .with_port_mapping(port_mapping)
        .with_control_server_address(control_port_addr)
        .finish()
        .await?;

let private_key = tor_hidden_service.private_key();
// Securely persist the private key which can be used to restore the onion address later.

    let comms_node =  CommsBuilder::new()
        .configure_from_hidden_service(&tor_hidden_service)
        // (...)
      .spawn()
      .await?;

```

## Motivation and Context

Closes #1100. Create a seamless tor integration for tari_comms.

## How Has This Been Tested?

Tested using the tor example - more fine grained tests are TODO, although the control port client library is extensively tested.

To run the example - start the tor proxy (ensuring that the `ControlPort 9051` setting is set in `torrc`) and run
`cargo run --example tor --features next -- /ip4/127.0.0.1/tcp/9051`

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/backoff.rs [hunk 1]
```diff
@@ -53,7 +53,7 @@ impl Default for ExponentialBackoff {
 
 impl Backoff for ExponentialBackoff {
     fn calculate_backoff(&self, attempts: usize) -> Duration {
-        if attempts == 0 {
+        if attempts <= 1 {
```
Comment (CjS77) [inline comms/src/backoff.rs]: Why this change?

Comment (sdbondi) [reply comms/src/backoff.rs]: The connection manager counts the first attempt as 1 which would add a 2 second delay instead of starting immediately.

Comment (sdbondi) [reply comms/src/backoff.rs]: The old connection manager (the one currently being used by most base node tests) starts at 0, making the first 2 attempts start immediately. Shouldn't be a problem, besides I'll be working on switching everything over to the new comms next 🎉

Comment (sdbondi) [reply comms/src/backoff.rs]: The nice thing about the new comms builder is that it allows you to specify the backoff - so for tests a `ConstantBackoff::new(Duration::from_secs(0))` might be more appropriate.
